### ISSUE TITLE
Annotation processing when using kapt results in compileJava failing with "no source files"

### ISSUE THREAD

**Issue by stoyicker** [gradle/gradle#8959]
### Expected Behavior
Compilation should succeed.

### Current Behavior
```
./gradlew clean :sample-kotlin:assemble spits an error:
* What went wrong:
Execution failed for task ':sample-kotlin:compileJava'.
> no source files
```

### Context
I am developing an annotation processor.

### Steps to Reproduce 
1. Get https://github.com/stoyicker/test-accessors/tree/2b5d420b7b0ae5a036d69cd4f85186d221e376f0
2. Run `./gradlew clean :sample-kotlin:assemble`

### Your Environment
Build scan URL: https://scans.gradle.com/s/o3642oxqlkj3k

**Comment by stoyicker** [comment]
This still reproduces with gradlew 6.6

**Comment by mlopatkin** [comment]
@stoyicker is the issue still reproducible with more recent versions of Gradle (7.x or 6.9)? Were you able to find a workaround?

**Comment by stoyicker** [comment]
Just tried with 7.1.1, still the same issue.

![log](https://user-images.githubusercontent.com/1732935/125443915-66c2b907-e88b-4fda-ba67-f8630a84e5f9.png)

https://gradle.com/s/kno4qext3arn4

**Comment by stoyicker** [comment]
Also, I didn't find any workarounds - instead just dropped support for Kotlin in favor of supporting newer versions of Gradle.

**Comment by mlopatkin** [comment]
I suspect that you've been hit by https://youtrack.jetbrains.com/issue/KT-31127. This issue should be fixed in Kotlin 1.3.50 and upgrading the Kotlin version in your reproducer makes the issue go away.

I'm closing this issue now because:
1. apparently the problem has been solved in the newer Kotlin plugin.
2. the issue is in the Kotlin plugin which is developed separately from the core Gradle so it cannot be addressed here.
